File: Univalued_Binary_Tree/solution.best.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
  def isUnivalTree(self, root: TreeNode) -> bool:
    return self._isUnivalTree(root, root)
  
  def _isUnivalTree(self, node: TreeNode, parent: TreeNode) -> bool:
    if node == None: return True
    if node.val != parent.val: return False
    if self._isUnivalTree(node.left, node) == False: return False
    return self._isUnivalTree(node.right, node)

File: Univalued_Binary_Tree/solution.best.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
  public boolean isUnivalTree(TreeNode root) {
    return this._isUnivalTree(root, root);
  }
  
  private boolean _isUnivalTree(TreeNode node, TreeNode parent) {
    if (node == null) return true;
    if (node.val != parent.val) return false;
    if (!this._isUnivalTree(node.left, node)) return false;
    return this._isUnivalTree(node.right, node);
  }
  
}
